Add additional performance tests

lucasponce opened the pull request:
This improvement will help to not block Java clients as metrics that send directly data into the engine.

jshaughn commented:
I agree that sendData should be more of a simple delivery of data with a fast return.  But lets hold on this PR until we can evaluate the incoming data flow a bit more, and see the effects of the change.  I don't think we have an issue here, so we can address this after a bit more study.

lucasponce commented:
I have reverted the asynchronous suggestion.
I also used this PR to add an additional performance test, it also complements other specific tests used from the engine and CRUD scenarios.
Added also some useful warning flags that will help to early determine if filtering/firing processing is under pressure.
I have been studying several scenarios to create some realistic perf tests that can help into perf investigations, I guess these ones are a step forward in that sense.

jshaughn commented on hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java:
```diff
@@ -155,20 +157,18 @@ public void fire() {
-        log.debug("firing rules...");
+        int initialPendingData = pendingData.size();
+        int initialPendingEvents = pendingEvents.size();
+        log.debugf("Firing rules... PendingData [%s] PendingEvents [%s]", initialPendingData, initialPendingEvents);
```
Shouldn't these 3 lines be under the 'while' below?

lucasponce replied:
initialPending* variables are used at the end of the computation to show final message.

jshaughn replied:
OK, then perhaps just the log message so we don't log on cycles with no pending data.

jshaughn commented on hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java:
```diff
@@ -50,6 +50,8 @@
     // private final MsgLogger msgLog = MsgLogger.LOGGER;
     private final Logger log = Logger.getLogger(DroolsRulesEngineImpl.class);
     private static final String SESSION_NAME = "hawkular-alerts-engine-session";
+    private static final long PERF_FILTERING_THRESHOLD= 3000L; // 3 seconds
```
We are using the term 'Filtering' to mean a few different things.  I think we should leave 'filtering' to mean rejecting data that is not used in conditions.  The pendingData is already 'filtered' and now we are ordering/deferring/batching it for submission to the engine.  Perhaps we can change this to PERF__BATCHING_THRESHOLD or something like that, and make analogous naming changes below?

lucasponce replied:
+1 to PERF_BATCHING_THRESHOLD

jshaughn commented on hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java:
```diff
@@ -182,6 +182,12 @@ public void fire() {
+            long filteringTime = System.currentTimeMillis() - startFiltering;
+            log.debugf("Filtering BatchData [%s] took [%s]", batchData.size(), filteringTime);
+            if (filteringTime > PERF_FILTERING_THRESHOLD) {
+                log.warnf("Filtering Filtering BatchData [%s] took [%s] ms exceeding [%s] ms",
```
Duplicate 'Filtering' in message (also see comment about changing 'Filtering' terminology).

Also, not show here in the diff, but above in the code we have this log message:
```
if (log.isDebugEnabled()) {
    log.debug("Deferring more recent " + data + " until older " + previousData + " is " + "processed");
}
```

Perhaps this should be removed completely.  For large amounts of data this may generate too much logging.  If we keep it, please convert to debugf.

lucasponce replied:
I guess the deferring info is really a large amount of data, on perf scenarios, it's not really interesting I think.
We take the assumption that the ordering is working fine.
I'm going to move them to tracef just in case.
It's harmless and it will be there just if needed.

lucasponce replied:
> Duplicate 'Filtering' in message (also see comment about changing 'Filtering' terminology).

Well, one is on debug and other in warning, content message is similar but not equals.

lucasponce replied:
Oh, I missed the "Filtering Filtering" string :)
Thanks for catching up, even after your review I have to read it twice :)

jshaughn commented on hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java:
```diff
@@ -220,6 +226,14 @@ public void fire() {
+            log.debugf("Firing took [%s] ms", firingTime);
+        }
+        if (firingTime > PERF_FIRING_THRESHOLD) {
+            log.warnf("Firing rules... PendingData [%s] PendingEvents [%s] took [%s] msexceeding [%s] ms",
```
'msexceeding' -> 'ms exceeding'

jshaughn commented on hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java:
```diff
@@ -155,36 +157,40 @@ public void fire() {
 
             pendingData.clear();
 
+            long startFiltering = System.currentTimeMillis();
```
If you don't mind, please change timing variables to 'startBatching' and 'batchingTime'.
